### DISCUSSION TITLE
Discrapency in metrics compared to kube-prometheus

### DISCUSSION THREAD

**Discussion by jethi** [coroot/coroot#241] [Q&A]
I am checking out Coroot in our Kubernetes cluster. We were using kube-prometheus for checking resource usage by pods. I noticed the CPU and Memory usage metrics for apps differ between Coroot and kube-prometheus.

I have a deployment running a single pod.

In Kubernetes the limit is set to 5Gi for the pod.
Coroot reports this correctly. 
Showing limit as 5.37 GB (5 Gi == 5.36871 GB rounded to 5.37 by Coroot)

When checking the current usage the story is weird.
The memory usage shown by kube-prometheus is 4.01 GiB
kubectl top pod <podname> reports 4099 Mi
So, kube-prometheus changes 4099Mi -> 4.01 GiB by a factor of 1/1000 seems to be the case.

Coroot reports usage as 3.65G

Now, I don't understand why reported usage is 4.01GiB in kube-prometheus top and 3.65G in coroot.

If I go by the factual difference reported in limits. Then the number is GiB should at least be lower than number reported in GB for the same value. But that is not the case.

So, I assume that kubectl top reports usage in Megabyte and not Mebibyte, Coroot reports it in Gibibyte instead of Gigabyte.
According to this assumption:

4099 MB is 4099 x 1000 x 1000 bytes

Then I try to convert it to GiB 4099000000/(1024 x 1024 x 1024) is roughly 3.817GiB

That is still off when compared to Coroot's reported usage of 3.65G

I am not able to establish a connection between the two. Please help.

**Comment by T100D** [comment]
The only difference I can think of is that prom gets the metrics from the interface properties in /proc/, and coroot gets the from eBPF.
